app/models/bulk_grant_permission_set.rb

```diff
@@ -0,0 +1,42 @@
+class BulkGrantPermissionSet < ActiveRecord::Base
+  belongs_to :user
+  has_many :bulk_grant_permission_set_application_permissions, inverse_of: :bulk_grant_permission_set
+  has_many :supported_permissions, through: :bulk_grant_permission_set_application_permissions
+
+  validates :user, presence: true
+  validates :outcome, inclusion: { in: %w(success fail), allow_nil: true }
+  validate :must_have_at_least_one_supported_permission
+
+
+  def in_progress?
+    outcome.nil?
+  end
+
+  def successful?
+    outcome == 'success'
+  end
+
+  def enqueue
+    BulkGrantPermissionSetJob.perform_later(self.id)
+  end
+
+  def perform(_options = {})
+    self.update_column(:total_users, User.count)
+    User.find_each do |user|
+      supported_permissions.each do |permission|
```

Comment:
Could we audit-log the permission changes here like we do for manual updates?

See https://github.com/alphagov/signon/pull/541:

https://github.com/alphagov/signon/blob/6208df339d20f0aa5c5bdb51f1b4b171d042f28c/app/services/user_update.rb#L34-L40

Reply:
oh, yes, absolutely.  I'll do a quick check to make sure that we're always adding event logs for all the places where we create users and grant permissions as I suspect this won't be the only place where we've slipped through the net.